taiga2/controllers/models_controller.py

```diff
@@ -736,48 +760,140 @@ def format_renamed(datafiles):
-    all_datafile_ids = [datafile.id for datafile in added_datafiles]
+    for index, folder in enumerate(parents):
+        if not can_view(folder.id):
+            del allowed_parents[index]
 
-    latest_dataset_version = get_latest_dataset_version(dataset_id)
+    return allowed_parents
 
-    datafile_diff = _get_dataset_version_datafiles_diff(
-        list(added_datafiles), list(latest_dataset_version.datafiles)
+
+def format_datafile_id(
+    dataset_permaname: str,
+    dataset_version: DatasetVersion,
+    datafile_name: Optional[str],
+):
+    name_parts = {
+        "dataset_permaname": dataset_permaname,
+        "dataset_version": dataset_version,
+        "datafile_name": datafile_name,
+    }
+
+    id_format = (
+        DATAFILE_ID_FORMAT
+        if datafile_name is not None
+        else DATAFILE_ID_FORMAT_MISSING_DATAFILE
+    )
+
+    return id_format.format(**name_parts)
+
+
+def get_previous_version_upload_datafiles(
+    dataset_version_datafiles: List[DataFile],
+    dataset_permaname: str,
+    dataset_version: str,
+) -> List[UploadVirtualDataFile]:
+    previous_version_taiga_ids = [
+        {
+            "taiga_id": format_datafile_id(
+                dataset_permaname, dataset_version, datafile.name
+            )
+        }
+        for datafile in dataset_version_datafiles
+    ]
+
+    previous_version_datafiles = (
+        [UploadVirtualDataFile(f) for f in previous_version_taiga_ids]
+        if previous_version_taiga_ids is not None
+        else None
     )
 
-    comments = _format_datafile_diff(datafile_diff)
+    return previous_version_datafiles
 
-    new_dataset_version = add_dataset_version(
-        dataset_id=dataset_id,
-        datafiles_ids=all_datafile_ids,
-        new_description=new_description,
-        changes_description=changes_description,
+
+def generate_previous_version_upload_files(existing_taiga_id, sid):
+    try:
+        data_file = get_datafile_by_taiga_id(existing_taiga_id, one_or_none=True)
+    except InvalidTaigaIdFormat as ex:
+        api_error(
+            "The following was not formatted like a valid taiga ID: {}".format(
+                ex.taiga_id
+            )
+        )
+
+    if data_file is None:
+        api_error("Unknown taiga ID: " + existing_taiga_id)
+
+    return add_upload_session_virtual_file(
+        session_id=sid, filename=data_file.name, data_file_id=data_file.id
     )
 
-    activity = VersionAdditionActivity(
-        user_id=current_user.id,
-        dataset_id=new_dataset_version.dataset_id,
-        type=Activity.ActivityType.added_version,
-        dataset_description=new_description
-        if new_description != latest_dataset_version.description
-        else None,
-        dataset_version=new_dataset_version.version,
-        comments=comments,
+
+def get_previous_version_and_added_datafiles(
+    dataset_version: DatasetVersion,
+    dataset_id: str,
+    datafile_names: List[str],
```

Comment:
This feels a little odd to me. I was trying to trace how this is being used, and it appears this parameter is just used to filter out old entries which conflict with names of new files.

However, the name of the function is `get_pervious_version_and_added_datafiles` so, it's surprising to me that it's also filtering out names. I think it'd be more natural to do the filtering outside of this function in the caller's body.

However, if you wanted to keep the filtering as-is, perhaps naming this parameter "datafile_names_to_exclude" would be more clear as to the intention.